// rsuite/esm/Uploader/Uploader.js
'use client';
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useCallback, useRef, useImperativeHandle, useReducer, useEffect } from 'react';
import find from 'lodash/find';
import FileItem from "./UploadFileItem.js";
import Plaintext from "../internals/Plaintext/index.js";
import ajaxUpload from "./utils/ajaxUpload.js";
import UploadTrigger from "./UploadTrigger.js";
import Box from "../internals/Box/index.js";
import { forwardRef, guid } from "../internals/utils/index.js";
import { useStyles, useCustom, useWillUnmount, useEventCallback } from "../internals/hooks/index.js";

// Define several states of the file during the upload process.

const getFiles = event => {
  if (typeof event?.['dataTransfer'] === 'object') {
    return event?.['dataTransfer']?.files;
  }
  if (event.target) {
    return event.target['files'];
  }
  return [];
};
const createFile = file => {
  const {
    fileKey
  } = file;
  return {
    ...file,
    fileKey: fileKey || guid(),
    progress: 0
  };
};
function fileListReducer(files, action) {
  switch (action.type) {
    // Add one or more files
    case 'push':
      return [...files, ...action.files];

    // Remove a file by `fileKey`
    case 'remove':
      return files.filter(f => f.fileKey !== action.fileKey);

    // Update a file
    case 'updateFile':
      return files.map(file => {
        return file.fileKey === action.file.fileKey ? action.file : file;
      });

    // Initialization file list
    case 'init':
      return action.files?.map(file => {
        // The state of the file needs to be preserved when the `fileList` is controlled.
        return files.find(f => f.fileKey === file.fileKey) || createFile(file);
      }) || [];
    default:
      throw new Error();
  }
}
const useFileList = (defaultFileList = []) => {
  const fileListRef = useRef(defaultFileList.map(createFile));
  const fileListUpdateCallback = useRef(null);
  const [fileList, dispatch] = useReducer(fileListReducer, fileListRef.current);
  fileListRef.current = fileList;
  useEffect(() => {
    fileListUpdateCallback.current?.(fileList);
    fileListUpdateCallback.current = null;
  }, [fileList]);
  useWillUnmount(() => {
    fileListUpdateCallback.current = null;
  });
  const dispatchCallback = useCallback((action, callback) => {
    dispatch(action);
    fileListUpdateCallback.current = callback;
  }, []);
  return [fileListRef, dispatchCallback];
};

/**
 * The `Uploader` component is used to upload files.
 *
 * @see https://rsuitejs.com/components/uploader
 */
const Uploader = forwardRef((props, ref) => {
  const {
    propsWithDefaults
  } = useCustom('Uploader', props);
  const {
    as,
    classPrefix = 'uploader',
    className,
    listType = 'text',
    defaultFileList,
    fileList: fileListProp,
    fileListVisible = true,
    locale,
    style,
    draggable,
    name = 'file',
    multiple = false,
    disabled = false,
    readOnly,
    plaintext,
    accept,
    children,
    toggleAs,
    removable = true,
    disabledFileItem,
    maxPreviewFileSize,
    method = 'POST',
    autoUpload = true,
    action,
    headers,
    withCredentials = false,
    disableMultipart,
    timeout = 0,
    data = {},
    onRemove,
    onUpload,
    shouldUpload,
    shouldQueueUpdate,
    renderFileInfo,
    renderThumbnail,
    onPreview,
    onChange,
    onSuccess,
    onError,
    onProgress,
    onReupload,
    onCompletion,
    ...rest
  } = propsWithDefaults;
  const {
    merge,
    withPrefix,
    prefix
  } = useStyles(classPrefix);
  const classes = merge(className, withPrefix());
  const rootRef = useRef(null);
  const xhrs = useRef({});
  const trigger = useRef(null);
  const uploadingCount = useRef(0);
  const uploadResults = useRef({
    completed: [],
    failed: []
  });
  const [fileList, dispatch] = useFileList(fileListProp || defaultFileList);
  useEffect(() => {
    if (typeof fileListProp !== 'undefined') {
      // Force reset fileList in reducer, when `fileListProp` is updated
      dispatch({
        type: 'init',
        files: fileListProp
      });
    }
  }, [dispatch, fileListProp]);
  const updateFileStatus = useCallback(nextFile => {
    dispatch({
      type: 'updateFile',
      file: nextFile
    });
  }, [dispatch]);

  /**
   * Clear the value in input.
   */
  const cleanInputValue = useCallback(() => {
    trigger.current?.clearInput();
  }, []);

  /**
   * Callback for successful file upload.
   * @param file
   * @param response
   * @param event
   * @param xhr
   */
  const handleAjaxUploadSuccess = useCallback((file, response, event, xhr) => {
    const nextFile = {
      ...file,
      status: 'finished',
      progress: 100
    };
    updateFileStatus(nextFile);
    onSuccess?.(response, nextFile, event, xhr);
    uploadingCount.current--;
    uploadResults.current.completed.push(nextFile);
    if (uploadingCount.current === 0) {
      onCompletion?.(uploadResults.current.completed, uploadResults.current.failed);
      uploadResults.current = {
        completed: [],
        failed: []
      };
    }
  }, [onCompletion, onSuccess, updateFileStatus]);

  /**
   * Callback for file upload error.
   * @param file
   * @param status
   * @param event
   * @param xhr
   */
  const handleAjaxUploadError = useCallback((file, status, event, xhr) => {
    const nextFile = {
      ...file,
      status: 'error'
    };
    updateFileStatus(nextFile);
    onError?.(status, nextFile, event, xhr);
    uploadingCount.current--;
    uploadResults.current.failed.push(nextFile);
    if (uploadingCount.current === 0) {
      onCompletion?.(uploadResults.current.completed, uploadResults.current.failed);
      uploadResults.current = {
        completed: [],
        failed: []
      };
    }
  }, [onCompletion, onError, updateFileStatus]);

  /**
   * Callback for file upload progress update.
   * @param file
   * @param percent
   * @param event
   * @param xhr
   */
  const handleAjaxUploadProgress = useCallback((file, percent, event, xhr) => {
    const nextFile = {
      ...file,
      status: 'uploading',
      progress: percent
    };
    updateFileStatus(nextFile);
    onProgress?.(percent, nextFile, event, xhr);
  }, [onProgress, updateFileStatus]);

  /**
   * Upload a single file.
   * @param file
   */
  const handleUploadFile = useEventCallback(file => {
    const {
      xhr,
      data: uploadData
    } = ajaxUpload({
      name,
      timeout,
      headers,
      data,
      method,
      withCredentials,
      disableMultipart,
      file: file.blobFile,
      url: action,
      onError: handleAjaxUploadError.bind(null, file),
      onSuccess: handleAjaxUploadSuccess.bind(null, file),
      onProgress: handleAjaxUploadProgress.bind(null, file)
    });
    updateFileStatus({
      ...file,
      status: 'uploading'
    });
    if (file.fileKey) {
      xhrs.current[file.fileKey] = xhr;
    }
    onUpload?.(file, uploadData, xhr);
  });
  const handleAjaxUpload = useEventCallback(() => {
    fileList.current.forEach(file => {
      const checkState = shouldUpload?.(file);
      if (checkState instanceof Promise) {
        uploadingCount.current++;
        checkState.then(res => {
          if (res) {
            handleUploadFile(file);
          } else {
            uploadingCount.current--;
            if (uploadingCount.current === 0) {
              onCompletion?.(uploadResults.current.completed, uploadResults.current.failed);
              uploadResults.current = {
                completed: [],
                failed: []
              };
            }
          }
        });
        return;
      } else if (checkState === false) {
        return;
      }
      if (file.status === 'inited') {
        uploadingCount.current++;
        handleUploadFile(file);
      }
    });
    cleanInputValue();
  });
  const handleUploadTriggerChange = useEventCallback(event => {
    const files = getFiles(event);
    const newFileList = [];
    Array.from(files).forEach(file => {
      newFileList.push({
        blobFile: file,
        name: file.name,
        status: 'inited',
        fileKey: guid()
      });
    });
    const nextFileList = [...fileList.current, ...newFileList];
    const checkState = shouldQueueUpdate?.(nextFileList, newFileList);
    if (checkState === false) {
      cleanInputValue();
      return;
    }
    const upload = () => {
      onChange?.(nextFileList, event);
      if (rootRef.current) {
        dispatch({
          type: 'push',
          files: newFileList
        }, () => {
          autoUpload && handleAjaxUpload();
        });
      }
    };
    if (checkState instanceof Promise) {
      checkState.then(res => {
        res && upload();
      });
      return;
    }
    upload();
  });
  const handleRemoveFile = useEventCallback((fileKey, event) => {
    const file = find(fileList.current, f => f.fileKey === fileKey);
    const nextFileList = fileList.current.filter(f => f.fileKey !== fileKey);
    if (xhrs.current?.[file.fileKey]?.readyState !== 4) {
      xhrs.current[file.fileKey]?.abort();
      uploadingCount.current--;
      uploadResults.current.failed.push(file);
      if (uploadingCount.current === 0) {
        onCompletion?.(uploadResults.current.completed, uploadResults.current.failed);
        uploadResults.current = {
          completed: [],
          failed: []
        };
      }
    }
    dispatch({
      type: 'remove',
      fileKey
    });
    onRemove?.(file);
    onChange?.(nextFileList, event);
    cleanInputValue();
  });
  const handleReupload = useEventCallback(file => {
    uploadingCount.current++;
    autoUpload && handleUploadFile(file);
    onReupload?.(file);
  });

  // public API
  const start = useCallback(file => {
    if (file) {
      uploadingCount.current++;
      handleUploadFile(file);
      return;
    }
    handleAjaxUpload();
  }, [handleAjaxUpload, handleUploadFile]);
  useImperativeHandle(ref, () => ({
    root: rootRef.current,
    start
  }));
  const renderList = [/*#__PURE__*/React.createElement(UploadTrigger, _extends({}, rest, {
    locale: locale,
    name: name,
    key: "trigger",
    multiple: multiple,
    draggable: draggable,
    disabled: disabled,
    readOnly: readOnly,
    accept: accept,
    ref: trigger,
    onChange: handleUploadTriggerChange,
    as: toggleAs
  }), children)];
  if (fileListVisible) {
    renderList.push(/*#__PURE__*/React.createElement("div", {
      key: "items",
      className: prefix('file-items')
    }, fileList.current.map((file, index) => /*#__PURE__*/React.createElement(FileItem, {
      locale: locale,
      key: file.fileKey || index,
      file: file,
      maxPreviewFileSize: maxPreviewFileSize,
      listType: listType,
      disabled: disabledFileItem,
      onPreview: onPreview,
      onReupload: handleReupload,
      onCancel: handleRemoveFile,
      renderFileInfo: renderFileInfo,
      renderThumbnail: renderThumbnail,
      removable: removable && !readOnly && !plaintext,
      allowReupload: !readOnly && !plaintext
    }))));
  }
  const dataAttributes = {
    'data-list-type': listType,
    'data-draggable': draggable
  };
  if (plaintext) {
    return /*#__PURE__*/React.createElement(Plaintext, _extends({
      localeKey: "notUploaded",
      className: classes
    }, dataAttributes), fileList.current.length ? renderList[1] : null);
  }
  if (listType === 'picture') {
    renderList.reverse();
  }
  return /*#__PURE__*/React.createElement(Box, _extends({
    as: as,
    ref: rootRef,
    className: classes,
    style: style
  }, dataAttributes), renderList);
});
Uploader.displayName = 'Uploader';
export default Uploader;
