// rsuite/esm/Animation/Transition.js
'use client';
import React from 'react';
import getTransitionEnd from 'dom-lib/getTransitionEnd';
import on from 'dom-lib/on';
import classNames from 'classnames';
import isFunction from 'lodash/isFunction';
import omit from 'lodash/omit';
import { getDOMNode } from "../internals/utils/index.js";
import { getAnimationEnd } from "./utils.js";
export let STATUS = /*#__PURE__*/function (STATUS) {
  STATUS[STATUS["UNMOUNTED"] = 0] = "UNMOUNTED";
  STATUS[STATUS["EXITED"] = 1] = "EXITED";
  STATUS[STATUS["ENTERING"] = 2] = "ENTERING";
  STATUS[STATUS["ENTERED"] = 3] = "ENTERED";
  STATUS[STATUS["EXITING"] = 4] = "EXITING";
  return STATUS;
}({});
const transitionProps = ['onEnter', 'onEntering', 'onEntered', 'onExit', 'onExiting', 'onExited', 'animation', 'children', 'className', 'in', 'unmountOnExit', 'transitionAppear', 'timeout', 'exitedClassName', 'exitingClassName', 'enteredClassName', 'enteringClassName'];

/**
 * A Transition component for animation.
 * @see https://rsuitejs.com/components/animation/#transition
 */
class Transition extends React.Component {
  static displayName = 'Transition';
  static defaultProps = {
    timeout: 1000
  };
  animationEventListener = null;
  instanceElement = null;
  nextCallback = null;
  needsUpdate = null;
  constructor(props) {
    super(props);
    let initialStatus;
    if (props.in) {
      initialStatus = props.transitionAppear ? STATUS.EXITED : STATUS.ENTERED;
    } else {
      initialStatus = props.unmountOnExit ? STATUS.UNMOUNTED : STATUS.EXITED;
    }
    this.state = {
      status: initialStatus
    };
    this.nextCallback = null;
    this.childRef = /*#__PURE__*/React.createRef();
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.in && nextProps.unmountOnExit) {
      if (prevState.status === STATUS.UNMOUNTED) {
        // Start enter transition in componentDidUpdate.
        return {
          status: STATUS.EXITED
        };
      }
    }
    return null;
  }
  getSnapshotBeforeUpdate() {
    if (!this.props.in || !this.props.unmountOnExit) {
      this.needsUpdate = true;
    }
    return null;
  }
  componentDidMount() {
    if (this.props.transitionAppear && this.props.in) {
      this.performEnter(this.props);
    }
  }
  componentDidUpdate() {
    const {
      status
    } = this.state;
    const {
      unmountOnExit
    } = this.props;
    if (unmountOnExit && status === STATUS.EXITED) {
      if (this.props.in) {
        this.performEnter(this.props);
      } else {
        if (this.instanceElement) {
          this.setState({
            status: STATUS.UNMOUNTED
          });
        }
      }
      return;
    }
    if (this.needsUpdate) {
      this.needsUpdate = false;
      if (this.props.in) {
        if (status === STATUS.EXITING || status === STATUS.EXITED) {
          this.performEnter(this.props);
        }
      } else if (status === STATUS.ENTERING || status === STATUS.ENTERED) {
        this.performExit(this.props);
      }
    }
  }
  componentWillUnmount() {
    this.cancelNextCallback();
    this.instanceElement = null;
  }
  onTransitionEnd(node, handler) {
    this.setNextCallback(handler);
    this.animationEventListener?.off();
    if (!this.nextCallback) {
      return;
    }
    if (node) {
      const {
        timeout,
        animation
      } = this.props;
      this.animationEventListener = on(node, animation ? getAnimationEnd() : getTransitionEnd(), this.nextCallback);
      if (timeout !== null) {
        setTimeout(this.nextCallback, timeout);
      }
    } else {
      setTimeout(this.nextCallback, 0);
    }
  }
  setNextCallback(callback) {
    let active = true;
    this.nextCallback = event => {
      if (!active) {
        return;
      }
      if (event) {
        if (this.instanceElement === event.target) {
          callback(event);
          active = false;
          this.nextCallback = null;
        }
        return;
      }
      callback(event);
      active = false;
      this.nextCallback = null;
    };
    if (this.nextCallback) {
      this.nextCallback.cancel = () => {
        active = false;
      };
    }
    return this.nextCallback;
  }
  getChildElement() {
    if (this.childRef.current) {
      return getDOMNode(this.childRef.current);
    }
    return getDOMNode(this);
  }
  performEnter(props) {
    const {
      onEnter,
      onEntering,
      onEntered
    } = props || this.props;
    this.cancelNextCallback();
    const node = this.getChildElement();
    this.instanceElement = node;
    onEnter?.(node);
    this.safeSetState({
      status: STATUS.ENTERING
    }, () => {
      onEntering?.(node);
      this.onTransitionEnd(node, () => {
        this.safeSetState({
          status: STATUS.ENTERED
        }, () => {
          onEntered?.(node);
        });
      });
    });
  }
  performExit(props) {
    const {
      onExit,
      onExiting,
      onExited
    } = props || this.props;
    this.cancelNextCallback();
    const node = this.getChildElement();
    this.instanceElement = node;
    onExit?.(node);
    this.safeSetState({
      status: STATUS.EXITING
    }, () => {
      onExiting?.(node);
      this.onTransitionEnd(node, () => {
        this.safeSetState({
          status: STATUS.EXITED
        }, () => {
          onExited?.(node);
        });
      });
    });
  }
  cancelNextCallback() {
    if (this.nextCallback !== null) {
      this.nextCallback.cancel();
      this.nextCallback = null;
    }
  }
  safeSetState(nextState, callback) {
    if (this.instanceElement) {
      const nextCallback = this.setNextCallback(callback);
      this.setState(nextState, () => nextCallback?.());
    }
  }
  render() {
    const status = this.state.status;
    if (status === STATUS.UNMOUNTED) {
      return null;
    }
    const {
      children,
      className,
      exitedClassName,
      enteringClassName,
      enteredClassName,
      exitingClassName,
      ...rest
    } = this.props;
    const childProps = omit(rest, transitionProps);
    let transitionClassName;
    if (status === STATUS.EXITED) {
      transitionClassName = exitedClassName;
    } else if (status === STATUS.ENTERING) {
      transitionClassName = enteringClassName;
    } else if (status === STATUS.ENTERED) {
      transitionClassName = enteredClassName;
    } else if (status === STATUS.EXITING) {
      transitionClassName = exitingClassName;
    }
    if (isFunction(children)) {
      childProps.className = classNames(className, transitionClassName);
      return children(childProps, this.childRef);
    }
    const child = React.Children.only(children);
    return /*#__PURE__*/React.cloneElement(child, {
      ...childProps,
      ref: this.childRef,
      className: classNames(className, child.props?.className, transitionClassName)
    });
  }
}
export default Transition;
