// rsuite/esm/InlineEdit/useEditState.js
'use client';
import { useState } from 'react';
import { useEventCallback, useControlled } from "../internals/hooks/index.js";
const useEditState = props => {
  const {
    value: valueProp,
    defaultValue,
    disabled,
    onChange,
    onEdit,
    onCancel,
    onSave,
    onClick,
    onFocus,
    ...htmlProps
  } = props;
  const [isEditing, setIsEditing] = useState(false);
  const [value, setValue] = useControlled(valueProp, defaultValue);

  // When editing, the value is not updated, and the original value is restored when canceling
  const [resetValue, setResetValue] = useState();
  const handleClick = useEventCallback(event => {
    if (disabled) {
      return;
    }
    onClick?.(event);
    onEdit?.(event);
    setIsEditing(true);
    setResetValue(value);
  });
  const handleFocus = useEventCallback(event => {
    if (disabled || isEditing) return;
    onFocus?.(event);
    setIsEditing(true);
    setResetValue(value);
  });
  const handleChange = useEventCallback((value, event) => {
    setValue(value);
    onChange?.(value, event);
  });
  const handleCancel = useEventCallback(event => {
    setIsEditing(false);
    setValue(resetValue);
    onCancel?.(event);
    event?.stopPropagation?.();
  });
  const handleSave = useEventCallback(event => {
    setIsEditing(false);
    onSave?.(event);
    event?.stopPropagation?.();
  });
  const handleKeyDown = useEventCallback(event => {
    if (isEditing) {
      switch (event.key) {
        case 'Enter':
          if (event.target?.tagName === 'INPUT') {
            handleSave(event);
          }
          break;
        case 'Escape':
          handleCancel(event);
          break;
      }
    }
  });
  return {
    isEditing,
    value,
    onClick: handleClick,
    onChange: handleChange,
    onFocus: handleFocus,
    onCancel: handleCancel,
    onSave: handleSave,
    onKeyDown: handleKeyDown,
    htmlProps
  };
};
export default useEditState;
