// rsuite/esm/Timeline/Timeline.js
'use client';
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import some from 'lodash/some';
import TimelineItem from "./TimelineItem.js";
import Box from "../internals/Box/index.js";
import { useStyles, useCustom } from "../internals/hooks/index.js";
import { forwardRef, rch } from "../internals/utils/index.js";
const ACTIVE_FIRST = index => index === 0;
const ACTIVE_LAST = (index, totalItemsCount) => index === totalItemsCount - 1;
const SubcomponentsAndStaticMethods = {
  Item: TimelineItem,
  ACTIVE_FIRST,
  ACTIVE_LAST
};

/**
 * The `Timeline` component is used to display a list of items in chronological order.
 *
 * @see https://rsuitejs.com/components/timeline
 */
const Timeline = forwardRef((props, ref) => {
  const {
    propsWithDefaults
  } = useCustom('Timeline', props);
  const {
    as = 'ul',
    children,
    classPrefix = 'timeline',
    className,
    align = 'left',
    endless,
    reverse,
    isItemActive = ACTIVE_LAST,
    ...rest
  } = propsWithDefaults;
  const {
    merge,
    withPrefix
  } = useStyles(classPrefix);
  const count = rch.count(children);
  const withTime = some(React.Children.toArray(children), item => item?.props?.time);
  const classes = merge(className, withPrefix(`align-${align}`, {
    endless,
    'with-time': withTime,
    reverse
  }));
  const childrenArray = React.Children.toArray(children);
  const orderedChildren = reverse ? [...childrenArray].reverse() : childrenArray;
  return /*#__PURE__*/React.createElement(Box, _extends({
    as: as,
    ref: ref,
    className: classes
  }, rest), orderedChildren.map((child, domIndex) => {
    const logicalIndex = reverse ? count - 1 - domIndex : domIndex;
    return /*#__PURE__*/React.cloneElement(child, {
      last: logicalIndex + 1 === count,
      INTERNAL_active: isItemActive(logicalIndex, count),
      align
    });
  }));
}, SubcomponentsAndStaticMethods);
Timeline.displayName = 'Timeline';
export default Timeline;
