// rsuite/cjs/InlineEdit/renderChildren.js
'use client';
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.defaultRenderInput = defaultRenderInput;
exports.renderChildren = renderChildren;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _Input = _interopRequireDefault(require("../Input"));
var _utils = require("../internals/utils");
function defaultRenderInput(props, ref) {
  return /*#__PURE__*/_react.default.createElement(_Input.default, (0, _extends2.default)({
    ref: ref
  }, props));
}
const pickers = ['DatePicker', 'DateRangePicker', 'TimePicker', 'TimeRangePicker', 'InputPicker', 'TagPicker', 'Cascader', 'MultiCascader', 'SelectPicker', 'CheckPicker', 'CheckTreePicker', 'TreePicker'];
function getDisplayName(Component) {
  if (typeof Component?.type === 'string') {
    return Component?.type;
  }
  return Component?.type?.displayName || '';
}
function renderChildren(children, props, ref) {
  if (typeof children === 'function') {
    return children(props, ref);
  }
  if (pickers.includes(getDisplayName(children))) {
    const {
      onBlur,
      ...rest
    } = props;
    const childElement = children;
    const {
      onExit,
      onClean
    } = childElement.props;
    return /*#__PURE__*/_react.default.cloneElement(childElement, {
      ...rest,
      // Pass onBlur to the child component to automatically save or cancel after the focus event is processed.
      // Special handling in the Picker component, call onBlur when onExit and onClean
      onExit: (0, _utils.createChainedFunction)(() => onBlur?.(), onExit),
      onClean: (0, _utils.createChainedFunction)(() => onBlur?.(), onClean),
      ref
    });
  }
  return /*#__PURE__*/_react.default.cloneElement(children, {
    ...props,
    ref
  });
}
